### PR TITLE
[gcs] Bootstrap dashboard with gcs address

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
This is part of redis removal. In this PR, we prefer gcs_address instead of getting it from redis. If gcs_address is set, it won't get gcs address from redis.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/agent.py [hunk 1]
```diff
@@ -158,9 +168,12 @@ async def _check_parent():
         # Start a grpc asyncio server.
         await self.server.start()
         # TODO: redis-removal bootstrap
```
Comment (mwtian) [inline dashboard/agent.py:170]: Remove this TODO?

File: dashboard/agent.py [hunk 2]
```diff
@@ -385,14 +399,16 @@ async def _check_parent():
...
             redis_client = ray._private.services.create_redis_client(
                 args.redis_address, password=args.redis_password)
-            gcs_publisher = None
             if args.gcs_address:
                 gcs_publisher = GcsPublisher(args.gcs_address)
+                redis_client = None
             elif gcs_pubsub_enabled():
                 gcs_publisher = GcsPublisher(
                     address=get_gcs_address_from_redis(redis_client))
+                redis_client = None
```
Comment (mwtian) [inline dashboard/agent.py:411]: How about initializing both gcs_publisher and redis_client to None, and setting redis_client only in an `else` branch here? An unnecessary initialization of redis_client can be avoided.

File: python/ray/worker.py [hunk 3]
```diff
@@ -602,6 +602,7 @@ def init(
         include_dashboard: Optional[bool] = None,
         dashboard_host: str = ray_constants.DEFAULT_DASHBOARD_IP,
         dashboard_port: Optional[int] = None,
+        gcs_server_port: Optional[int] = None,
```
Comment (mwtian) [inline python/ray/worker.py:605]: Can this be prefixed with `_`? I wonder why there is no such flag for Redis port. And adding option to ray.init() may require more review.

Comment (iycheng) [reply python/ray/worker.py:605]: should this be prefixed? i'm not sure about the rule here. i can update this.

**Comment by ericl** [comment]
Worker tests failing with redis error

File: python/ray/tests/conftest.py [hunk 4]
```diff
@@ -52,6 +52,8 @@ def get_default_fixture_ray_kwargs():
 def _ray_start(**kwargs):
     init_kwargs = get_default_fixture_ray_kwargs()
     init_kwargs.update(kwargs)
+    if os.environ.get("RAY_bootstrap_with_gcs") in ("true", "1"):
+        init_kwargs["gcs_server_port"] = 9993
```
Comment (ericl) [inline python/ray/tests/conftest.py:56]: Any reason not to use the (previous) redis port?

File: python/ray/node.py [hunk 5]
```diff
@@ -745,12 +745,18 @@ def start_dashboard(self, require_dashboard):
                 if we fail to start the dashboard. Otherwise it will print
                 a warning if we fail to start the dashboard.
         """
+        gcs_address = None
+        if self._ray_params.gcs_server_port is not None:
```
Comment (ericl) [inline python/ray/node.py:749]: Can we add TODOs to remove these blocks when redis is removed, here and elsewhere?

**Review by ericl — COMMENTED** [review]
It would be good for readability to clearly mark what parts of code are legacy/should be removed, since this is adding quite a lot of branches.

File: python/ray/tests/conftest.py (hunk 4 above)
Comment (iycheng) [reply python/ray/tests/conftest.py:56]: redis port by default is that port. so I'm thinking maybe gcs shouldn't use that. but for test i can update it to that.

**Comment by iycheng** [comment]
i'll open with a new one. close this for now.

**Pull request closed without merging**